cmd/kubeadm/app/util/staticpod/utils.go

```diff
@@ -252,15 +252,15 @@ func GetAPIServerProbeAddress(endpoint *kubeadmapi.APIEndpoint) string {
 
 // GetControllerManagerProbeAddress returns the kubernetes controller manager probe address
 func GetControllerManagerProbeAddress(cfg *kubeadmapi.ClusterConfiguration) string {
-	if addr, exists := cfg.ControllerManager.ExtraArgs[kubeControllerManagerAddressArg]; exists {
+	if addr, exists := cfg.ControllerManager.ExtraArgs[kubeControllerManagerBindAddressArg]; exists {
```

Comment:
will a probe address of `0.0.0.0` work?

Reply:
good catch 
```
--- FAIL: TestGetControllerManagerProbeAddress (0.00s)
    --- FAIL: TestGetControllerManagerProbeAddress/setting_controller_manager_extra_ipv6_address_arg_to_something_acknowledges_it#01 (0.00s)
        /Users/aojea/work/src/k8s.io/kubernetes/cmd/kubeadm/app/util/staticpod/utils_test.go:141: Unexpected result from GetControllerManagerProbeAddress:
            	expected: 127.0.0.1
            	actual: 0.0.0.0
```

Reply:
I guess that we should default to localhost for unspecified addresses, right?
my question is, should we return `localhost` or try to be smart and resolve the ip family?

Reply:
@mauilion I remember that you told me that in linux is possible to use 0.0.0.0 as destination address, but do you think that will be safe to use it to configure the http probe?

Reply:
it feels to me that if `0.0.0.0` is passed as `--bind-address` it should be used for a probe too, transparently, without defaulting to a localhost or per family IP.

Reply:
I think that we should fall back to loopbacks addresses for the probes, it's the most secure and reliable IMHO

Reply:
alright.

Reply:
I mean, if it binds to 0.0.0.0 is going to listen in all interfaces and all ip addresses ... but I don't understand what IP is going to poll the probe, is deterministic the 0.0.0.0 or it just picks one randomly in the host?

Reply:
> I mean, if it binds to 0.0.0.0 is going to listen in all interfaces and all ip addresses ... but I don't understand what IP is going to poll the probe, is deterministic the 0.0.0.0 or it just picks one randomly in the host?

I have the same question... is contacting 0.0.0.0 a normal thing to do?

Reply:
we mostly care for the probes to serve on a loopback IP, yet the 0.0.0.0 -> loopback IP transform becomes a undocumented behavior.

if someone complains they can customize their manifests.

Reply:
does a probe against 0.0.0.0 actually work?

Reply:
_checking_ in the meantime net.Dial considers it valid :/
https://golang.org/src/net/dial.go

Reply:
One thing I'm worried about here is whether using 0.0.0.0 for the probe works now because of obscure reasons but will stop working in the future; I'm not sure if this is legal at all. Also, would this work on kubelet on Windows?

Reply:
`kind create cluster`
kubectl apply
```
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-http
spec:
  containers:
  - name: liveness
    image: k8s.gcr.io/liveness
    args:
    - /server
    livenessProbe:
      httpGet:
        host: 0.0.0.0
        path: /healthz
        port: 8080
        httpHeaders:
        - name: Custom-Header
          value: Awesome
      initialDelaySeconds: 3
      periodSeconds: 3
```

and the winner is 😄 

```
Events:
  Type     Reason     Age                 From                         Message
  ----     ------     ----                ----                         -------
  Normal   Scheduled  109s                default-scheduler            Successfully assigned default/liveness-http to kind-control-plane
  Normal   Pulled     73s (x3 over 100s)  kubelet, kind-control-plane  Successfully pulled image "k8s.gcr.io/liveness"
  Normal   Created    72s (x3 over 99s)   kubelet, kind-control-plane  Created container liveness
  Normal   Started    72s (x3 over 99s)   kubelet, kind-control-plane  Started container liveness
  Normal   Pulling    63s (x4 over 108s)  kubelet, kind-control-plane  Pulling image "k8s.gcr.io/liveness"
  Warning  Unhealthy  63s (x9 over 96s)   kubelet, kind-control-plane  Liveness probe failed: Get http://0.0.0.0:8080/healthz: dial tcp 0.0.0.0:8080: connect: connection refused
  Normal   Killing    63s (x3 over 90s)   kubelet, kind-control-plane  Container liveness failed liveness probe, will be restarted
```

Reply:
I've created another one without the host parameter and it works

the curious is that the url without host works
```
    Liveness:       http-get http://:8080/healthz delay=3s timeout=1s period=3s #success=1 #failure=3
```
but this doesn't
```
    Liveness:       http-get http://0.0.0.0:8080/healthz delay=3s timeout=1s period=3s #success=1 #failure=3
```

Reply:
ok, in that case our fallback here is valid.

Reply:
> the curious is that the url without host works

this could be defaulting that k8s probes are doing.

Reply:
using 0.0.0.0 will end in a broken environment hard to debug (it took me a while today to find out the problem) and is common to use those addresses, at least we use `::` in kind , I think that in case of having bind-address as 0.0.0.0 or :: we should choose one of the following options:
1. omit the host field in the probe
2. use a loopback address

Reply:
I'm leaning towards omitting the host field, at the end 0.0.0.0 and :: is kind of "no address" :)

Reply:
So looks like omitting the port is the best way to go, so we don't need ad-hoc logic for ipv4 and ipv6 here, right?

Reply:
kubeadm itself does block 0.0.0.0 at least as an api server advertise address: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/config/common.go#L121-L140
but not for custom bind-address (extra args) for the other components.

we also have to note that the api server probe uses the advertise address, while the probes for the other components use bind-address
https://github.com/kubernetes/kubernetes/blob/34ba1683b9d68284a481814d58673f999cff0520/cmd/kubeadm/app/util/staticpod/utils.go#L237-L251

so it seems to me a api server probe of 0.0.0.0 was already not possible.

Reply:
For the sake of completeness: https://tools.ietf.org/html/rfc5735#section-3.

Reply:
Commented on #86504